### DISCUSSION TITLE
Moving from Flask to FastAPI

### DISCUSSION THREAD

**Discussion by mshkanth** [AgnostiqHQ/covalent#852] [General]
**Description**
Various discussions between Team Psi & Team AQ have led to a unanimous decision to move from flask to FastAPI in the Covalent core & web app components, in the short or medium term. This discussion is to align on why, what & how of this porting. 

**Impacted components**
1. Backend APIs of the web app (maintained by Team Psi) - As a part of the DB revamp exercise, all new back end APIs have been written in FastAPI. 
2. Covalent core backend APIs (maintained by Team AQ)
3. Socket implementation between Web app back end & web app GUI (maintained by Team Psi)
4. Serving of UI files from the back end (maintained by Team Psi)

**Impacted files**

1.  `covalent_ui/app.py`
a. Contains few old backend APIs that won't work with the new DB schema & need to be de-commissioned. 
b. Flask initialization code that runs APIs. 
c. Serving of UI files

2. `covalent_dispatcher/_service/app.py` - Contains few covalent core APIs written using Flask. 

**Roadmap for each impacted compoent**

1. `covalent_ui/app.py`

It is proposed to de-commission/remove the following APIs as they aren't needed for the GUI & won't work with the new DB schema. These APIs have been re-factored into a set of smaller & different API end points to be consumed by the GUI - 
a. `/api/results`
b. `/api/dev/results/<dispatch_id>`
c. `/api/results/<dispatch_id>`
d. `/api/logoutput/<dispatch_id>`

It is proposed to migrate the following APIs to FastAPI - 
a. `/api/webhook`
b. `/api/draw`

2. A socket implementation compatible with FastAPI will be implemented in place of the current `flask_socketio` implementation. In addition, the task of serving the UI will also be moved to FastAPI to maintain a clean exit from Flask. Additionally, it is proposed to keep this app.py lean and have minimal wrapper FastAPI code. Only initialization & first level configurations will be done in this file. The implementation of API routes & other details will be done outside this file and injected into the FastAPI instance initiated from here. 

_Note : All of the above changes are being undertaken by Team Psi as part of the DB revamp excercise._ 

3. `covalent_dispatcher/_service/app.py`

~~It is proposed that Team AQ takes responsibility of ideation & execution of porting to FastAPI.~~ Team AQ have requested for a single team to take ownership of the FastAPI migration. Team Psi will also take up migration of the above file. It is suggested that a single FastAPI instance initialized in `covalent_ui/app.py` be used as the hook for both Covalent core & Covalent web app APIs when they are ported to FastAPI.

**Pending decisions**

~~Team Psi is waiting for Team AQ to come back with a plan of action for `covalent_dispatcher/_service/app.py`.~~ Team Psi will also take up migration of the above file.

**Comment by mshkanth** [comment]
@wjcunningham7 @santoshkumarradha - For your kind attention

**Reply by mshkanth** [reply]
@wjcunningham7 @santoshkumarradha - Team Psi is taking up the migration of the Flask aspects of the Covalent dispatcher. The AQ side components are highlighted above. Pls let us know if we have missed anything.

**Comment by mshkanth** [comment]
@wjcunningham7  @AlejandroEsquivel - `covalent_dispatcher/_service/app.py` creates instances of `ThreadPoolExecutor()` and passes it via `dispatch_id = dispatcher.run_dispatcher(json_lattice, workflow_pool, tasks_pool)`. As we port the `@bp.route("/submit", methods=["POST"])` route to FastAPI, do you anticipate issues with retaining ThreadPoolExecutor() in the FastAPI implementation?

Wanted to share this issue for your reference - https://github.com/tiangolo/fastapi/issues/3316

**Reply by santoshkumarradha** [reply]
Is this taken care off ? @wjcunningham7 ?

**Reply by mshkanth** [reply]
@santoshkumarradha - We haven't seen any performance drops with this implementation. However I will let Will/Alejandro comment.

**Reply by AlejandroEsquivel** [reply]
As far as I understand I don't see an issue, @kessler-frost any thoughts?

**Reply by kessler-frost** [reply]
I don't anticipate any issues with retaining `ThreadPoolExecutor` for now. Eventually we would want to get rid of them, but that will only be possible once we have completely switched to FastAPI.

Short answer: No objections.

**Reply by mshkanth** [reply]
Hi @kessler-frost - We have indeed moved completely to FastAPI across Dispatcher APIs & Web app APIs. No components run on Flask now. Whenever you feel this is worthwhile to look into, pls highlight / reach out.

**Reply by kessler-frost** [reply]
Hi @mshkanth that's great! We've started to look into the logical changes required and I'll reach out if there's any issues. Thanks!
